### PR TITLE
[BB-3853] Prepare Ocim for mass upgrade of instances to Koa

### PR DESCRIPTION
**Pull request by 0x29a**
This PR contains various changes related to switching to the new Open edX release:
1. https://github.com/open-craft/opencraft/pull/682, but updated to use `koa.2a` release. Also, [integration-koa](https://github.com/open-craft/configuration/tree/integration-koa) branch is rebased on [open-release/koa.2a](https://github.com/edx/configuration/tree/open-release/koa.2a).
1. Updated default Open edX release. New beta-test instances will use `koa.2a` by default.
1. Updates to the markdown files.
1. Updated API client code to reflect changes in the `help_text`.
1. Updated Ubuntu image settings, and migrations to update default DB values based on these settings.

**JIRA tickets**:
- [BB-3853](https://tasks.opencraft.com/browse/BB-3853).

**Dependencies**:
- This PR depends on https://github.com/open-craft/ansible-secrets/pull/249.

**Testing instructions**:

### I: CI
Check that PR passed all tests and checks in Circle CI.

### II: Prepare stage
1. Connect to the stage Ocim: `ssh ubuntu@stage.manage.opencraft.com`.
2. Attach to the screen session: `screen -r 1384.stage`.
3. Change window to `bash`.
4. Fetch PR branch: `git fetch origin 0x29a/bb3853/prepare_ocim_for_koa`.
5. Checkout to the PR branch: `git checkout 0x29a/bb3853/prepare_ocim_for_koa`.
6. Open `.env` file and change:
    - Value of `OPENSTACK_SANDBOX_BASE_IMAGE` to `'{"name": "focal-20.04-unmodified"}'`.
    - Value of `OPENEDX_RELEASE_STABLE_REF` to `open-release/koa.2a`.
    - Value of `STABLE_CONFIGURATION_VERSION` to `opencraft-release/koa.2a`.
    - Value of `STABLE_EDX_PLATFORM_COMMIT` to `opencraft-release/koa.2a`.
    - Value of `SIMPLE_THEME_SKELETON_THEME_VERSION` to `samuel/remove-v2`.
7. Change screen window to `ipython`, **restart shell**.
8. Make sure that:
   - `settings.OPENSTACK_SANDBOX_BASE_IMAGE` is equal to `'{"name": "focal-20.04-unmodified"}'`.
   - `settings.OPENEDX_RELEASE_STABLE_REF` is equal to `open-release/koa.2a`.
   - `settings.STABLE_CONFIGURATION_VERSION` is equal to `opencraft-release/koa.2a`.
   - `settings.STABLE_EDX_PLATFORM_COMMIT` is equal to `opencraft-release/koa.2a`.
   - `settings.SIMPLE_THEME_SKELETON_THEME_VERSION` is equal to `samuel/remove-v2`.
   We need to do this because migrations fetch default values from settings.
9. Change screen window again to `bash`.
10. Run `make migrate`.
11. Change screen window to `web`, hit `CTRL + C`, run `make run`.

### III: Beta-test instance
1. Open https://stage.console.opencraft.com/ in private browser window and register trial account with `name+bb3853@opencraft.com` email address.
2. Go to https://stage.manage.opencraft.com/ and make sure that Ocim created beta-test instance.
3. Open its settings and make sure that:
   - `openedx_release` is equal to `open-release/koa.2a`.
   - `openstack_server_base_image` is equal to `'{"name": "focal-20.04-unmodified"}'`.
   - `configuration_version` is equal to `opencraft-release/koa.2a`.
   - `edx_platform_commit` is equal to `opencraft-release/koa.2a`.
4. Wait for instance to finish app server provisioning.
5. [Optionally] Perform [manual testing](https://gitlab.com/opencraft/documentation/private/-/blob/master/checklists/manual_instance_test.md) of the created instance.

### IV: Cleanup
1. Change screen window to `bash`.
2. `honcho run python3 manage.py migrate pr_watch 0013_add_ansible_overrides`.
3. `honcho run python3 manage.py migrate instance 0139_auto_20210217_1216`.
4. Open `.env` file and change:
    - Value of `OPENSTACK_SANDBOX_BASE_IMAGE` to `'{"name": "xenial-16.04-unmodified"}'`.
    - Value of `OPENEDX_RELEASE_STABLE_REF` to `open-release/juniper.3`.
    - Value of `STABLE_CONFIGURATION_VERSION` to `opencraft-release/juniper.3`.
    - Value of `STABLE_EDX_PLATFORM_COMMIT` to `opencraft-release/juniper.3`.
    - Value of `SIMPLE_THEME_SKELETON_THEME_VERSION` to `master`.
5. `git checkout master`.
6. Change screen window to `ipython`, **restart shell**.
7. Change screen window to `web`, hit `CTRL + C`, run `make run`.

**Author notes and concerns**:
- After merging this and _right before_ deploying this on prod, we have to merge https://github.com/open-craft/edx-simple-theme/pull/16 **and** remove [this line](https://github.com/open-craft/opencraft/blob/f2325310a24feb38deb7258f3420324b0116d16e/circle.yml#L153).

**Reviewers**
- [ ] @lgp171188

### CODE REVIEW

File: circle.yml
Comment (lgp171188) [inline circle.yml:153]: @0x29a, this PR branch should be merged to a properly named branch and that should be used here. :) If that branch is `master`, this should be removed. :)

File: documentation/configuration.md
Comment (lgp171188) [inline documentation/configuration.md:79]: @0x29a, it is worth mentioning that this is the case from Koa onwards.

File: frontend/packages/api_client/src/models/DisplayStaticContentPages.ts
Comment (lgp171188) [inline frontend/packages/api_client/src/models/DisplayStaticContentPages.ts:57]: @0x29a, how is this and the other changes in this file related to the Koa upgrade? 🤔

File: frontend/packages/api_client/src/models/OpenEdXAppServer.ts
Comment (lgp171188) [inline frontend/packages/api_client/src/models/OpenEdXAppServer.ts:110]: @0x29a, is there any reason to expose these fields via the API in the first place?

File: instance/migrations/0140_auto_20210315_1246.py
Comment (lgp171188) [inline instance/migrations/0140_auto_20210315_1246.py:1]: @0x29a: nit: use a proper name for the migration file that explains what the change is.

File: instance/models/mixins/openedx_config.py
Comment (lgp171188) [inline instance/models/mixins/openedx_config.py:491]: @0x29a, there is no `forum_version` in Koa, just `FORUM_VERSION`.

File: instance/models/mixins/openedx_config.py
Comment (lgp171188) [inline instance/models/mixins/openedx_config.py:489]: @0x29a, can you also add some details about the code drift/commits that require us to use the fork?

File: instance/models/openedx_appserver.py
Comment (lgp171188) [inline instance/models/openedx_appserver.py:166]: @0x29a, since this change triggers an additional migration, can we just leave it as-is. It is just the help text. Same for the 2 other similar changes.

File: opencraft/settings.py
Comment (lgp171188) [inline opencraft/settings.py:325]: @0x29a, shouldn't this be

```
{"name": "focal-20.04-unmodified"}
```
?
Let's use the same default values everywhere.

File: playbooks/collect_activity/stats.py
Comment (lgp171188) [inline playbooks/collect_activity/stats.py:1]: @0x29a, why is this change needed? Doesn't `/edx/bin/python.edxapp` use the same virtualenv environment?

File: instance/models/mixins/openedx_config.py
Comment (lgp171188) [reply instance/models/mixins/openedx_config.py:489]: @0x29a, also since we have no instances on Ironwood and it is not supported anyways, can we remove the if condition for it and the other ones above?

This function is becoming becoming messy and we will have to clean it up properly if we want to have release-specific configuration for multiple unsupported releases.

File: pr_watch/migrations/0014_auto_20210315_1246.py
Comment (lgp171188) [inline pr_watch/migrations/0014_auto_20210315_1246.py:1]: @0x29a, nit: name the migration file properly to indicate the change that it applies.

**Review by lgp171188 — COMMENTED** [review]
@0x29a, I left some comments. Please check those out.

File: circle.yml
Comment (0x29a) [reply circle.yml:153]: @lgp171188, I covered this in `Author notes and concerns` section of the PR description, but I think we have to merge this branch in `master` after merging this PR because doing this _before_ might break customization of non-Koa trial instances.

File: frontend/packages/api_client/src/models/DisplayStaticContentPages.ts
Comment (0x29a) [reply frontend/packages/api_client/src/models/DisplayStaticContentPages.ts:57]: @lgp171188, we don't modify this file manually. I just ran [update-api-client.sh](https://github.com/open-craft/opencraft/blob/master/frontend/scripts/update-api-client.sh) script to reflect changes in the `help_text` (it fetches API schema using Swagger endpoing). Looks like someone has forgotten to do this after changing backend code.

File: frontend/packages/api_client/src/models/OpenEdXAppServer.ts
Comment (0x29a) [reply frontend/packages/api_client/src/models/OpenEdXAppServer.ts:110]: @lgp171188, yep, we need to expose these fields to show their values in the Ocim interface (`Settings` tab):
image

Comment (lgp171188) [reply frontend/packages/api_client/src/models/OpenEdXAppServer.ts:110]: @0x29a. that doesn't use the new frontend APIs written in TS that this file is about, right? The settings tab in question uses the old APIs the clients for which are in written in Angular, no?

File: instance/models/mixins/openedx_config.py
Comment (0x29a) [reply instance/models/mixins/openedx_config.py:489]: @lgp171188, we have periodic builder for Ironwood: https://manage.opencraft.com/instance/19890/.

File: instance/models/openedx_appserver.py
Comment (0x29a) [reply instance/models/openedx_appserver.py:166]: @lgp171188, not only this change triggers migrations. Note that the default value for this field is calculated lazily basing on settings. I included these migrations to update default values. During deployment, we'll have to check that `settings.OPENSTACK_SANDBOX_BASE_IMAGE` (and other settings, I mentioned all of them in testing instructions) have correct values.

File: playbooks/collect_activity/stats.py
Comment (0x29a) [reply playbooks/collect_activity/stats.py:1]: @lgp171188, AFAIK `python.edxapp` [doesn't exist](https://discuss.openedx.org/t/koa-installation-attempt-missing-python-edxapp/3872/2?u=0x29a) in Koa. [Original Shimul's commit](https://github.com/open-craft/opencraft/commit/07b7857dce0974b0af56dcfd01100005d384e92f).

File: frontend/packages/api_client/src/models/OpenEdXAppServer.ts
Comment (0x29a) [reply frontend/packages/api_client/src/models/OpenEdXAppServer.ts:110]: @lgp171188 looks like both API versions use the [same serializer](https://github.com/open-craft/opencraft/blob/f99a37607c1d9ca45512091b72eb48fb1c170189/instance/serializers/openedx_instance.py#L93-L187). Should I do something about it?

File: instance/models/mixins/openedx_config.py
Comment (lgp171188) [reply instance/models/mixins/openedx_config.py:489]: @0x29a, wouldn't it be simpler and easier to just add the configuration override to that one instance?

Comment (0x29a) [reply instance/models/mixins/openedx_config.py:489]: You right, @lgp171188. By the way, we have `hawthorn.1` instance: https://manage.opencraft.com/instance/13740/

Refactoring this would require testing the deployment for both instances. Would it make sense to create a follow-up ticket instead of doing this withing this PR?

**Comment by 0x29a** [comment]
It's :green_circle:, @lgp171188. :slightly_smiling_face:

File: frontend/packages/api_client/src/models/OpenEdXAppServer.ts
Comment (lgp171188) [reply frontend/packages/api_client/src/models/OpenEdXAppServer.ts:110]: @demid, let's leave it as-is.

File: instance/models/mixins/openedx_config.py
Comment (lgp171188) [reply instance/models/mixins/openedx_config.py:490]: @0x29a, the issue with follow-up tasks is that they will get pushed out and deprioritized for various reasons and might get lost. And then we will end up with ugly code that is probably not needed any more. Can we just add the configuration overrides to the affected instances and add a note not to redeploy them and proceed with the upgrade? Checking and fixing issues, if any, on those tasks can be a follow-up task.

File: instance/models/openedx_appserver.py
Comment (lgp171188) [reply instance/models/openedx_appserver.py:166]: @0x29a, defining default values for a model based on a setting is a very bad practice. A callable should have been used instead :( But let's not worry about this now.

File: instance/models/mixins/openedx_config.py
Comment (0x29a) [reply instance/models/mixins/openedx_config.py:490]: Makes sense, @lgp171188. I will do this shortly.

**Review by lgp171188 — APPROVED** [review]
@0x29a, :+1: Apart from the one open discussion thread these changes look okay to me. Do you think you can refactor the changes into multiple logical commits before merging? It's okay if that will be time-consuming.


- [x] I **did not** test this on stage because it is being used for testing various other changes and there is not much time left before we have to complete the upgrades. We can and will have to fix issues, if any as soon as they come up during the upgrade. But the code changes look okay to me, So I am approving based on that.
- [x] I read through the code
- [x] ~~I checked for accessibility issues~~ NA
- [x] ~~Includes documentation~~ NA

File: instance/models/mixins/openedx_config.py
Comment (0x29a) [reply instance/models/mixins/openedx_config.py:472]: Done, @lgp171188. Here is an extra configuration I added for the Hawthorn instance:
```yaml
# master and juniper release onwards has djangoapps.heartbeat installed by default
# openedx <= ironwood requires this if celery check is included in
# heartbeat extended checks (which we add by default above)
EDXAPP_CMS_ENV_EXTRA:
  ADDL_INSTALLED_APPS: [ 'openedx.core.djangoapps.heartbeat' ]
EDXAPP_LMS_ENV_EXTRA:
  ADDL_INSTALLED_APPS: [ 'openedx.core.djangoapps.heartbeat' ]
  HEARTBEAT_EXTENDED_CHECKS:
    - lms.lib.comment_client.utils.check_forum_heartbeat
```

For Ironwood:
```yaml
# master and juniper release onwards has djangoapps.heartbeat installed by default
# openedx <= ironwood requires this if celery check is included in
# heartbeat extended checks (which we add by default above)
EDXAPP_CMS_ENV_EXTRA:
  ADDL_INSTALLED_APPS: [ 'openedx.core.djangoapps.heartbeat' ]
EDXAPP_LMS_ENV_EXTRA:
  ADDL_INSTALLED_APPS: [ 'openedx.core.djangoapps.heartbeat' ]
  HEARTBEAT_EXTENDED_CHECKS:
    - lms.lib.comment_client.utils.check_forum_heartbeat

# See https://github.com/edx/cs_comments_service/pull/323 and https://github.com/open-craft/cs_comments_service/pull/5
forum_source_repo: https://github.com/open-craft/cs_comments_service.git
forum_version: opencraft-release/ironwood.2
```

**Pull request merged**
